ext/spl/spl_fixedarray.c

```diff
@@ -150,7 +150,13 @@ static void spl_fixedarray_dtor_range(spl_fixedarray *array, zend_long from, zen
 {
 	zval *begin = array->elements + from, *end = array->elements + to;
 	while (begin != end) {
-		zval_ptr_dtor(begin++);
+		if (Z_REFCOUNTED_P(begin)) {
```

Comment:
Alright, CI has a nice red color. And it's because of this if check.
The optimizer const-evaluates `$obj[2] = str_repeat('B', 10);` (from the test code) into a constant string, so `Z_REFCOUNTED_P` is false. This means we get a different result when opcache is on vs when it's off.

I can just get rid of this if check and the problem goes away. This means that all previous elements will be set to NULL, even if they're primitives like int or bool or whatever.
However that might be a BC break, because you never know that there's people accessing previous elements (non-refcounted) in their destructors.
If anyone has an opinion on this please let me know :-)

Reply:
Or, also a BC break but may be cleaner: set the size prior to calling the destructors such that accesses to destroyed items result in an invalid index exception.

Reply:
I believe that with two phases you can still trigger the same error if you're *really* trying. E.g if you have an object a with a reference to b, and both are in the array, a might have a refcount of 1 and b of 2, thus calling the destructor only for a but not for b. That means the destructor of b would only be called in the second phase, again allowing you to observe the intermediate state of the array.

I feel like just moving `ZVAL_NULL(begin)` out of the `if` is a good solution. Any such code is highly questionable to begin with.

Reply:
RE: two phases. Icky.

I actually found another problem that wouldn't be solved with the nullifying, but would be with forcing the invalid index error by setting the size prior to calling the destructors. I have added this as a separate commit (will of course squash upon merging).